site/src/pages/_app.tsx

```diff
@@ -107,6 +101,61 @@ const amplitude = localFont({
   ],
   variable: "--site-font-family-amplitude",
 });
+const useHasHydrated = () => {
+  const [hasHydrated, setHasHydrated] = useState(false);
+
+  useEffect(() => {
+    setHasHydrated(true);
+  }, []);
+
+  return hasHydrated;
+};
+
+interface ThemeProviderProps {
+  /** Applies to `SaltProvider` `theme` prop */
+  themeClassName?: string;
+  className?: string;
+  children?: ReactNode;
+  /** Enables Salt theme next */
+  themeNext?: boolean;
+}
+
+// This is a direct copy of Mosaic's ThemeProvider + injecting density, so that we can control top level provider's density,
+// which impacts both children as well as portal
+function ThemeProvider({
+  themeClassName,
+  className,
+  themeNext,
+  children,
+}: ThemeProviderProps) {
+  const hasHydrated = useHasHydrated();
+  const colorMode = useColorMode();
+
+  const ssrClassname = hasHydrated ? undefined : ssrClassName;
+
+  const ChosenSaltProvider = themeNext
+    ? UNSTABLE_SaltProviderNext
+    : SaltProvider;
+  const viewport = useCurrentBreakpoint();
```

Comment:
I don't think this is working, the default site always returns touch

Reply:
Good catch, it's to do with hook on the same level of provider. Fixed with a workaround.